### PR TITLE
Debian moves scls to the core package

### LINKED ISSUE #2978: Debian packaging shouldn't require mod-elastic for elasticsearch-http module
## Description of the problem
Use of elasticsearch-http is only contingent upon the http module, and not the presence of the elasticsearch2 or java modules.  Requiring mod-elastic to use the elasticsearch-http module ends up pulling in the Java framework unnecessarily.

## Proposed solution
As there are two distinct files in scl/ for the two ES connection modules, retain elasticsearch2 (scl/elasticsearch/plugin.conf) for mod-elastic, and include elasticsearch-http (scl/elasticsearch/elasticsearch-http.conf) for mod-http.

### PR DESCRIPTION
**Pull request by bazsi**
This should fix #2978

### CODE REVIEW

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by bazsi** [comment]
Build packages please

**Comment by kira-syslogng** [comment]
Package Build SUCCESS

Packages are available at https://ci.syslog-ng.com/jenkins/job/kira-dbld-pull-requests/614/artifact/

File: scl/scl.conf [hunk 1]
```diff
@@ -25,6 +25,8 @@
 # `include-path`, then includes all SCL supplied plugins.
 #
 
+@module appmodel
```
Comment (Kokan) [inline scl/scl.conf]: Could you please explain why this line should be added ?

**Comment by bazsi** [comment]
Because scls use application adapters, which needs to be available. An
alternative would be to use @requires in all affected scls.

I didn't want to use requires in the main config because that would skip
the entire scl.conf.

Also, if we were to use @requires that would silently skip application
adapters, which is not intuitive behavior.

I was testing this patchset by disabling automatic loading of modules, and
that revealed this case.

Hth,
Bazsi

**Comment by lbudai** [comment]
@bazsi : hmm... I remember that we've talked about this a few months ago, but still don't really like this approach. I think packages should be self contained.

**Comment by lbudai** [comment]
@bazsi:
If I understand, the main motivation behind moving SCLs into syslog-ng-core is to somehow limit package proliferation.

(I think if we would have an SCL package manager tool then we could pre-install SCLs.
This tool could hide OS specific package manager tools and we could write this:
scl install http
scl depends elastic

Or, we can add an @auto-install-deps directive to the config which then calls this tool to install the dependencies... - a kind of 'lazy install').


As we don't have this tool, I'd say that when a module is implemented as an SCL and
 * it doesn't have any dependencies, it should be part of the core module
 * if it has one and only one dependency (a syslog-ng-mod-xxx), it should be part of that package
 * if it has more than one dependency, it should be a separate package that depends on the others

This is a clear model from dependency management point of view, but yes, it is true: there could be a huge number of packages that contains only a single SCL (but the key point here is that this SCL has dependencies).

On the other hand, if Debian maintainers switch to the model you suggested, then I think we should merge this change into syslog-ng.


So the right question is maybe who is responsible for handling dependencies for an SCL module? Currently it has a dual nature... 
 * SCL defines dependencies itself (via `@requires`)
 * and OS packages also define the very same dependencies ...

I hope that we can discuss these questions in IRL next week.

**Comment by bazsi** [comment]
I don't think the current state / size of the scl would warrant splitting
it up into packages. A few lines installed in the core package should not
be  a significant overhead, and significantly smaller then the dedicated
packages would be (the size of the package information used by apt/dpkg
would be larger than the scl file).

Also, syslog-ng should give you proper error message in case the scl is
installed and the underlying module is not. If the error message is not
good enough let's improve those, instead of splitting up scl into gazillion
packages, each with a few lines of syslog-ng configuration in them.

My expectation is that we would have hundreds of files in scl. Keeping
those in separate packages simply doesn't scale.

And the storage overhead of the entire library is negligible.

We might at one point reach a state where a specific part of the scl
becomes huge while not all users need that (for instance the application
adapters stuff), in that case I think it would make sense to package those
bits separately. But wrappers around java/python does not warrant separate
packages in my opinion.

Bazsi

**Comment by lbudai** [comment]
Trying to summarize and move things forward.

### What are SCLs for?
syslog-ng config language (or configuration library) used for implement new 'element' (parser, source, destination, etc...) by creating a 'config block' from existing 'modules'.

Sometimes these 'modules' are depending on a generic purpose module, like elasticsearch-http destination which is a smart http module.

#### Packaging
The rule what we tried to follow in the past:
 * 1 module : 1 OS package
 * when a module does not have any other dependencies, goes into syslog-ng-core, otherwise goes into a separate package

#### What about http-destination based modules (SCLs)?
 * some of them are come with syslog-ng-mod-http package (telegram or slack destinations)
 * some of them are moved to the corresponding java-based package (elasticsearch, kafka)

None of them is really good... especially the second one (users cannot install non-java based drivers without installing java).

The n module - 1 package could be work, but ... well, it solves only one (mainly theoretical) issue (-> self contained packages), while users has to know which package installs what modules.

But there is another problem: there are SCLs that could be part of more than one package. This breaks the rule, so need something else.

A possible quick solution is to move all SCLs into syslog-ng-core package (this is what this PR does).

As a consequence if there is a library which is reponsible for parsing a grammar and there is an SCL which produces a config block(parser, dst, src,...) should be parsed by this grammar then the library and the SCL are deployed by two separate package. What does this mean? There is a contract between SCL and grammar which is an issue only when someone doesn't want to upgrade a specific module to a newer version while upgrades syslog-ng-core.
I don't think this is a real problem, just wanted to show an example why I'm resisting a little bit against this change: this hurts self-contained modules (of course, the level of self-containment can also be challenged when the modules are all depending on libsyslog-ng...).

'Not great, not terrible'...  so I'm approving this.

**Comment by bazsi** [comment]
> This only solves #2978 if users will use syslog-ng `@requires` system instead of OS packaging.
> What I mean, shown with the elasticsearch example:
> If the user installs syslog-ng from package management (syslog-ng package), then subsequently he would like to add the elasticsearch **driver**, he looks for a package named `syslog-ng.*elastic`.
> But if he installs it, `elasticsearch-http()` still won't work, he has to install `syslog-ng-mod-http`.

This is not this simple, as we are not considering a "static" system, rather a configuration file that the user is about to start syslog-ng with.

If the "elastic-http.conf" SCL module is not available (not installed, or in this example, not included), then this is what you get if you start syslog-ng:

```
Error parsing destination statement, destination plugin elasticsearch-http not found in etc/syslog-ng-elastic.conf:7:16-7:34:
2       
3       #@include "scl/elasticsearch/elastic-http.conf"
4       
5       log {
6       	source { tcp(port(2000)); };
7-----> 	destination { elasticsearch-http(index("syslog") type("syslog") url("http://localhost:1234/")); };
7-----> 	              ^^^^^^^^^^^^^^^^^^
8       };
```

e.g. syslog-ng is complaining about the missing elasticsearch-http() plugin.

On the other hand, when I remove the http module, this remains the same, as the @requires will completely skip the inclusion of `elastic-http.conf`, even if it's present.

But if I push `@requires` into the block, this could start complaining about the missing http() plugin, but if I do that the lexer crashes for some reason, that is non-trivial to fix.

All in all, if we can get syslog-ng to complain about the missing http() plugin when someone tries to use the elasticsearch-http() block, then we are better off than not installing `elastic-http.conf` in the first place.


> 
> I guess this was the argument against having **every** SCLs in core on today's discussuin, as package maintainers don't like this approach, if I understand it right.
> It's currently inconsistent, you are right @bazsi, but this solution has caveats too.
> 
> Shouldn't we check what are the distro Maintainers' requirements/ practices?
> 
> I've just quickly checked, but elasticsearch is not included in the main packages on Ubuntu 19.04.
> https://packages.ubuntu.com/source/disco/syslog-ng
> And I have not found it on OpenSuse Tumbleweed either:
> https://software.opensuse.org//download.html?project=openSUSE%3ALeap%3A15.1%3AUpdate&package=syslog-ng

I think the reason here is that it would depend on java and none of our java based stuff is included in distros. Probably because we don't support the latest java versions and also because our java dependenciesT (e.g. the kafka client libs, elasticsearch etc) might not be packaged.

> 
> Or what about @lbudai's suggestion (and AFAIK current way of packaging):
> 
> > As we don't have this tool, I'd say that when a module is implemented as an SCL and
> > it doesn't have any dependencies, it should be part of the core module
> > if it has one and only one dependency (a syslog-ng-mod-xxx), it should be part of that package
> > if it has more than one dependency, it should be a separate package that depends on the others

This would potentially cause a lot of SCL-only packages to be created causing a maintenance headache and potential mistakes made when/how to create a new package for a couple of lines of SCL.

File: scl/elasticsearch/elastic-java.conf [hunk 2]
```diff
@@ -21,6 +21,7 @@
 #############################################################################
 
 @requires json-plugin
```
[the remark this answers is not in the document]
Comment (bazsi) [reply scl/elasticsearch/elastic-java.conf]: the java version needs it too as payload tends to be json formatted and the default template is `$(format-json)` based too.

**Comment by Kokan** [comment]
@kira-syslogng retest this please

**Comment by kira-syslogng** [comment]
Build FAILURE

File: scl/elasticsearch/elastic-java.conf (hunk 2 above)
[the remark this answers is not in the document]
Comment (gaborznagy) [reply scl/elasticsearch/elastic-java.conf]: Sorry, I meant add it to elastic-http too.
I've added it in https://github.com/bazsi/syslog-ng/pull/3

**Comment by gaborznagy** [comment]
@bazsi I was very curious if an underscore/hyphen error caused kira to fail, and when I found out it is, I thought why not just send a patch and save you some time.

Also, I've added an additional patch that elasticsearch-http() should require json-plugin too.

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by gaborznagy** [comment]
Wow, first kira-automake failed https://ci.syslog-ng.com/jenkins/job/kira-starter/4580/,
now kira-cmake https://ci.syslog-ng.com/jenkins/job/kira-starter/4586/.
Strange, it's the same failure, however I've tested it, even on top of @bazsi's branch...

I leave this as it is for now ...

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build FAILURE

**Comment by kira-syslogng** [comment]
Build SUCCESS

**Comment by bazsi** [comment]
I think there's nothing this should not be merged, is there? I have also created a branch that
improves error diagnostics in case of dependent modules #3015 so both could go in in the same
release.

**Pull request merged**